File: pirates-android/app/src/main/java/com/crescentflare/piratesgame/page/modules/splash/SplashLoaderModule.kt
package com.crescentflare.piratesgame.page.modules.splash

import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.content.Context
import android.view.View
import android.view.animation.AccelerateDecelerateInterpolator
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.binder.InflatorMapBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import com.crescentflare.piratesgame.components.compoundviews.SplashLoadingBar
import com.crescentflare.piratesgame.infrastructure.events.AppEvent
import com.crescentflare.piratesgame.infrastructure.events.AppEventObserver
import com.crescentflare.piratesgame.page.storage.Page
import com.crescentflare.piratesgame.page.modules.ControllerModule
import kotlinx.coroutines.*
import java.lang.ref.WeakReference

/**
 * Splash module: handles loading and updating the loading bar component
 */
class SplashLoaderModule: ControllerModule {

    // --
    // Static: inflatable integration
    // --

    companion object {

        val inflatable: JsonInflatable = object : JsonInflatable {

            override fun create(context: Context): Any {
                val module = SplashLoaderModule()
                module.onCreate(context)
                return module
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                return obj is SplashLoaderModule
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj is SplashLoaderModule
            }

        }

    }


    // --
    // Members
    // --

    override val eventType = "splashLoader"
    private var context: WeakReference<Context>? = null
    private var binder: InflatorMapBinder? = null
    private var loadingTasks = mutableListOf<LoadingTask>()
    private var busy = false
    private var preparing = false
    private var done = false


    // --
    // Lifecycle
    // --

    override fun onCreate(context: Context) {
        // Keep weak reference to context
        this.context = WeakReference(context)

        // Add loading tasks
        loadingTasks.add(LoadingTask {
            // Template for a loading task
        })
    }

    override fun onResume() {
        // No implementation
    }

    override fun onPause() {
        // No implementation
    }


    // --
    // Page updates
    // --

    override fun onPageUpdated(page: Page, binder: InflatorMapBinder) {
        this.binder = binder
        if (busy) {
            showLoading(false)
            if (done) {
                (binder.findByReference("loadingBar") as? SplashLoadingBar)?.progress = 1f
            }
        }
    }


    // --
    // Event handling
    // --

    override fun catchEvent(event: AppEvent, sender: Any?): Boolean {
        if (event.rawType == eventType) {
            if (!preparing) {
                preparing = true
                startLoading()
                return true
            }
        }
        return false
    }


    // --
    // Loading
    // --

    private fun startLoading() {
        GlobalScope.launch(Dispatchers.Main) {
            delay(1000)
            busy = true
            showLoading()
            startLoadTask {
                GlobalScope.launch(Dispatchers.Main) {
                    val bar = binder?.findByReference("loadingBar") as? SplashLoadingBar
                    if (bar?.autoAnimation == true) {
                        delay(SplashLoadingBar.animationDuration)
                    }
                    val eventObserver = context?.get() as? AppEventObserver
                    eventObserver?.observedEvent(AppEvent("navigate://summary?type=replace"), null)
                    done = true
                }
            }
        }
    }

    private fun startLoadTask(index: Int = 0, completion: () -> Unit) {
        if (context?.get() == null) {
            return
        }
        if (index >= loadingTasks.size) {
            completion()
            return
        }
        loadingTasks[index].start {
            val bar = binder?.findByReference("loadingBar") as? SplashLoadingBar
            bar?.progress = (index + 1).toFloat() / loadingTasks.size.toFloat()
            startLoadTask(index + 1, completion)
        }
    }

    private fun showLoading(animated: Boolean = true) {
        val bar = binder?.findByReference("loadingBar") as? View
        val text = binder?.findByReference("loadingText") as? View
        bar?.visibility = View.VISIBLE
        text?.visibility = View.VISIBLE
        if (animated) {
            if (bar != null && text != null) {
                val animation = AnimatorSet()
                bar.alpha = 0f
                text.alpha = 0f
                animation.playTogether(
                    ObjectAnimator.ofFloat(bar, View.ALPHA, 0f, 1f),
                    ObjectAnimator.ofFloat(text, View.ALPHA, 0f, 1f)
                )
                animation.duration = 250
                animation.interpolator = AccelerateDecelerateInterpolator()
                animation.start()
            }
        } else {
            bar?.alpha = 1f
            text?.alpha = 1f
        }
    }


    // --
    // Loading task
    // --

    private class LoadingTask(private val delay: Long = 0, private val loader: () -> Unit) {

        fun start(completion: () -> Unit) {
            GlobalScope.launch(Dispatchers.Default) {
                if (delay > 0) {
                    delay(delay)
                }
                loader()
                GlobalScope.launch(Dispatchers.Main) {
                    completion()
                }
            }
        }

    }

}

File: pirates-ios/PiratesGame/Page/Modules/Splash/SplashLoaderModule.swift
//
//  SplashLoaderModule.swift
//  Shared module: catches alert events to show popup dialogs
//

import UIKit
import UniLayout
import JsonInflator

fileprivate class SplashLoaderModuleTask {
    
    private let delay: Double
    private let loader: () -> Void
    
    init(delay: Double = 0, loader: @escaping () -> Void) {
        self.delay = delay
        self.loader = loader
    }
    
    func start(completion: @escaping () -> Void) {
        DispatchQueue.global().asyncAfter(deadline: .now() + delay, execute: {
            self.loader()
            DispatchQueue.main.async {
                completion()
            }
        })
    }
    
}

class SplashLoaderModule: ControllerModule {
    
    // --
    // MARK: Members
    // --
    
    let eventType = "splashLoader"
    private weak var viewController: UIViewController?
    private var binder: InflatorDictBinder?
    private var loadingTasks = [SplashLoaderModuleTask]()
    private var busy = false
    private var preparing = false
    private var done = false

    
    // --
    // MARK: Inflator integration
    // --
    
    class func inflatable() -> JsonInflatable {
        return InflatorClass()
    }
    
    private class InflatorClass: JsonInflatable {
        
        func create() -> Any {
            return SplashLoaderModule()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            return object is SplashLoaderModule
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return object is SplashLoaderModule
        }
        
    }


    // --
    // MARK: Initialization
    // --
    
    init() {
        // No implementation
    }


    // --
    // MARK: Lifecycle
    // --

    func didCreate(viewController: UIViewController) {
        // Keep reference to the view controller
        self.viewController = viewController
        
        // Add loading tasks
        loadingTasks.append(SplashLoaderModuleTask(loader: {
            // Template for a loading task
        }))
    }
    
    func didPause() {
        // No implementation
    }
    
    func didResume() {
        // No implementation
    }

    
    // --
    // MARK: Page updates
    // --

    func didUpdatePage(page: Page, binder: InflatorDictBinder) {
        self.binder = binder
        if (busy) {
            showLoading(animated: false)
            if (done) {
                (binder.findByReference("loadingBar") as? SplashLoadingBar)?.progress = 1.0
            }
        }
    }


    // --
    // MARK: Event handling
    // --

    func catchEvent(_ event: AppEvent, sender: Any?) -> Bool {
        if event.rawType == eventType {
            if !preparing {
                preparing = true
                startLoading()
                return true
            }
        }
        return false
    }


    // --
    // MARK: Loading
    // --
    
    private func startLoading() {
        DispatchQueue.main.asyncAfter(deadline: .now() + 1, execute: {
            self.busy = true
            self.showLoading()
            self.startLoadTask {
                var delay: Double = 0
                let bar = self.binder?.findByReference("loadingBar") as? SplashLoadingBar
                if bar?.autoAnimation == true {
                    delay = SplashLoadingBar.animationDuration
                }
                DispatchQueue.main.asyncAfter(deadline: .now() + delay, execute: {
                    let eventObserver = self.viewController as? AppEventObserver
                    eventObserver?.observedEvent(AppEvent(string: "navigate://summary?type=replace"), sender: nil)
                    self.done = true
                })
            }
        })
    }
    
    private func startLoadTask(index: Int = 0, completion: @escaping () -> Void) {
        if viewController == nil {
            return
        }
        if index >= loadingTasks.count {
            completion()
            return
        }
        loadingTasks[index].start {
            let bar = self.binder?.findByReference("loadingBar") as? SplashLoadingBar
            bar?.progress = Float(index + 1) / Float(self.loadingTasks.count)
            self.startLoadTask(index: index + 1, completion: completion)
        }
    }
    
    private func showLoading(animated: Bool = true) {
        let bar = binder?.findByReference("loadingBar") as? SplashLoadingBar
        let text = binder?.findByReference("loadingText") as? UniTextView
        bar?.visibility = .visible
        text?.visibility = .visible
        if animated {
            if let bar = bar, let text = text {
                bar.alpha = 0
                text.alpha = 0
                UIView.animate(withDuration: 0.25, delay: 0.0, options: [.curveEaseInOut], animations: {
                    bar.alpha = 1
                    text.alpha = 1
                })
            }
        } else {
            bar?.alpha = 1.0
            text?.alpha = 1.0
        }
    }

}
